compiler/grainformat/reformat.re

```diff
@@ -3308,258 +4127,177 @@ let print_primitive_value_description =
-                ),
+    | PTopExportAll(export_excepts) =>
+      Doc.concat([
+        Doc.text("export * "),
+        if (List.length(export_excepts) > 0) {
+          Doc.concat([
+            Doc.text("except "),
+            Doc.join(
+              Doc.comma,
+              List.map(
+                (excpt: Parsetree.export_except) =>
+                  switch (excpt) {
+                  | ExportExceptData(data) => Doc.text(data.txt)
+                  | ExportExceptValue(value) => Doc.text(value.txt)
+                  },
+                export_excepts,
               ),
-            ]);
-          } else {
-            Doc.nil;
-          },
-        ])
-      };
+            ),
+          ]);
+        } else {
+          Doc.nil;
+        },
+      ])
+    };
 
-    Doc.concat([
-      stmt_leading_comment_docs,
-      blank_line_above,
-      attribute_text,
-      Doc.group(without_comments),
-      line_end_comments,
-    ]);
-  };
+  Doc.group(without_comments);
 };
 
 let reformat_ast =
     (
+      ~original_source: array(string),
       parsed_program: Parsetree.parsed_program,
-      original_source: array(string),
     ) => {
-  let _ =
-    Walktree.walktree(parsed_program.statements, parsed_program.comments);
+  let get_loc = (stmt: Parsetree.toplevel_stmt) => {
+    stmt.ptop_loc;
+  };
 
-  let last_stmt = ref(None);
-  let previous_line = ref(0);
-  let printed_doc =
-    Doc.concat([
-      Doc.join(
-        Doc.hardLine,
-        List.map(
-          (stmt: Grain_parsing.Parsetree.toplevel_stmt) => {
-            let line =
-              toplevel_print(
-                ~data=stmt,
-                ~original_source,
-                ~previous_line=previous_line^,
-              );
+  let print_item = (stmt: Parsetree.toplevel_stmt, comments) => {
+    toplevel_print(~original_source, ~comments, stmt);
+  };
 
-            previous_line := get_end_loc_line(stmt.ptop_loc);
-            last_stmt := Some(stmt);
-            line;
-          },
-          parsed_program.statements,
-        ),
-      ),
-    ]);
+  let leading_comments = [];
 
-  let (_leading_comments, trailing_comments) =
-    switch (last_stmt^) {
-    | None => ([], [])
-    | Some(stmt) =>
-      Walktree.partition_comments(
-        ~range=None,
-        ~leading_only=false,
-        stmt.ptop_loc,
-      )
-    };
+  let cleaned_comments =
+    remove_used_comments(
+      ~all_comments=parsed_program.comments,
+      leading_comments,
+    );
 
-  let trailing_comment_docs =
-    if (List.length(trailing_comments) > 0) {
-      Doc.concat([print_multi_comments(trailing_comments, previous_line^)]);
-    } else {
-      Doc.nil;
-    };
+  let top_level_stmts =
+    block_item_iterator(
+      ~previous=TopOfFile,
+      ~get_loc,
+      ~print_item,
+      ~comments=cleaned_comments,
+      ~get_attribute_text=
+        stmt => {
+          let attributes = stmt.ptop_attributes;
+          print_attributes(attributes);
+        },
```

Comment:
create a `let` binding for this function?

Reply:
Can you explain why that's better?  Is it a performance thing?

Reply:
Nope, just a personal style nit. I don't like the look of anonymous functions passed as a named argument to a function.

Reply:
Changed.